plugins/camera/camera_impl.cpp

```diff
@@ -0,0 +1,739 @@
+#include "camera_impl.h"
+#include "camera_definition.h"
+#include "device.h"
+#include "global_include.h"
+#include "mavlink_include.h"
+#include "http_loader.h"
+#include <functional>
+
+namespace dronecore {
+
+using namespace std::placeholders; // for `_1`
+
+CameraImpl::CameraImpl(Device &device) :
+    PluginImplBase(device)
+{
+    _parent.register_plugin(this);
+}
+
+CameraImpl::~CameraImpl()
+{
+    _parent.unregister_plugin(this);
+}
+
+void CameraImpl::init()
+{
+    _parent.register_mavlink_message_handler(
+        MAVLINK_MSG_ID_CAMERA_CAPTURE_STATUS,
+        std::bind(&CameraImpl::process_camera_capture_status, this, _1),
+        this);
+
+    _parent.register_mavlink_message_handler(
+        MAVLINK_MSG_ID_STORAGE_INFORMATION,
+        std::bind(&CameraImpl::process_storage_information, this, _1),
+        this);
+
+    _parent.register_mavlink_message_handler(
+        MAVLINK_MSG_ID_CAMERA_IMAGE_CAPTURED,
+        std::bind(&CameraImpl::process_camera_image_captured, this, _1),
+        this);
+
+    _parent.register_mavlink_message_handler(
+        MAVLINK_MSG_ID_CAMERA_SETTINGS,
+        std::bind(&CameraImpl::process_camera_settings, this, _1),
+        this);
+
+    _parent.register_mavlink_message_handler(
+        MAVLINK_MSG_ID_CAMERA_INFORMATION,
+        std::bind(&CameraImpl::process_camera_information, this, _1),
+        this);
+
+    _parent.send_command_with_ack_async(
+        MAV_CMD_REQUEST_CAMERA_INFORMATION,
+        MavlinkCommands::Params {1.0f, // request it
+                                 NAN, NAN, NAN, NAN, NAN, NAN},
+        nullptr,
+        MAV_COMP_ID_CAMERA);
+}
+
+void CameraImpl::deinit()
+{
+    delete _camera_definition;
+
+    _parent.unregister_all_mavlink_message_handlers(this);
+}
+
+void CameraImpl::enable() {}
+
+void CameraImpl::disable() {}
+
+Camera::Result CameraImpl::take_photo()
+{
+    std::lock_guard<std::mutex> lock(_capture.mutex);
+
+    return camera_result_from_command_result(
```

Comment:
Don't we need to make sure we're in Photo mode? Same applies to async versions.

Reply:
Yap, thx.